src/v/archival/probe.cc

```diff
@@ -166,20 +166,33 @@ void ntp_level_probe::setup_dynamic_public_metrics(const model::ntp& ntp) {
           .aggregate(aggregate_labels),
         sm::make_gauge(
           "records_pending_upload",
-          [this] { return _parent.records_pending_upload(); },
+          [this] { return records_pending_upload(); },
           sm::description("The number of records in the local log that have "
                           "not yet been uploaded"),
           labels)
           .aggregate(aggregate_labels),
         sm::make_gauge(
           "estimated_backlog_size_bytes",
-          [this] { return _parent.estimate_backlog_size(); },
+          [this] { return _parent.last_estimated_backlog_size(); },
           sm::description("Estimated upload backlog size in bytes"),
           labels)
           .aggregate(aggregate_labels),
       });
 }
 
+int64_t ntp_level_probe::records_pending_upload() {
+    if (
+      !_last_records_pending_refresh
+      || ss::lowres_clock::now()
+           > _last_records_pending_refresh.value()
+               + _records_pending_refresh_interval
+               + _records_pending_refresh_jitter.next_duration()) {
+        _records_pending_upload = _parent.records_pending_upload();
```

Comment:
This kind of on-read caching will still result in hitting the slow code during the prometheus scrape, just not on every scrape. 

Maybe tag a last_records_pending_upload onto the ntp_archiver, and refresh it as part of estimated_backlog_size() since that's already called regularly?  (with some level of commenting/renaming functions to make it clear who's doing what)

Reply:
Sure. Updated the PR with your proposal.